database/rrdcalc.c

```diff
@@ -570,7 +577,10 @@ void rrdcalc_unlink_and_free(RRDHOST *host, RRDCALC *rc) {
     debug(D_HEALTH, "Health removing alarm '%s.%s' of host '%s'", rc->chart?rc->chart:"NOCHART", rc->name, host->hostname);
 
     // unlink it from RRDSET
-    if(rc->rrdset) rrdsetcalc_unlink(rc);
+    struct rrdcalc_rrdset_alarm search;
+    search.st = rc->rrdset;
+    struct rrdcalc_rrdset_alarm *rra = (struct rrdcalc_rrdset_alarm *)avl_search_lock(&host->alarms_idx_health_name, (avl *)&search);
```

Comment:
You should move those lines inside `rrdsetcalc_unlink()` and not change the function prototype. You can get the `rrdset` from inside the function, in the first line where it does `RRDSET *st = rc->rrdset;`.

Reply:
Are you doing this for performance reasons? To save 2 lookups?

Reply:
Yes, it was due performance.